src/library_fs.js

```diff
@@ -2021,6 +2021,15 @@ FS.staticInit();` +
         transaction.onerror = onerror;
       };
       openRequest.onerror = onerror;
+    },
+
+    // Allocate memory for an mmap operation. This allocates space of the right
+    // page-aligned size, and clears it.
+    mmapAlloc: function(size) {
+      size = alignMemory(size, {{{ POSIX_PAGE_SIZE }}});
+      var ptr = _malloc(size);
+      Array.prototype.fill.call(HEAP8, 0, ptr, ptr + size)
```

Comment:
This is quite slow for two reasons:

1. We zero the full memory, even though it would be sufficient to zero out the last page (actually just the padding bytes).
2. `Array#fill()` on a typed array is not particularly fast in today's JSVMs. it'd be better to just have a handrolled loop here.

WDYT?

Reply:
Huh, it's not fast? Yes, I did some benchmarking now and indeed it isn't, that's disappointing...

A loop is much better, and typed array fill much better still, so let's ifdef over those two I guess, #11570

Reply:
About point (1), that would take some extra refactoring to move the zeroing around to the specific FSes as opposed to a single shared place, which is a little unfortunate.

Meanwhile I measure that typed array fill can do 100MB of zeroing in ~10ms on my machine, which seems to indicate further optimization may not be worth it?

Reply:
We actually only care about v8 here, since we're working on a Chrome extension. Why not simply always use the loop though? (the #ifdef's look pretty terrible to me TBH).

100MB in 10ms is quite a bummer given that we currently read 1GB via mmap in ~16ms. So we'd add 100ms to that for nothing. If you remember original and aligned size, you can just fill from original size to aligned size, that should be really fast.

Reply:
The loop is 10x slower than typed array fill. But yes, if we want to just do the alignment area, I can look into a larger refactoring for that, which avoids the speed issue entirely.

Reply:
To avoid the refactor can we just zero the final page?   (I assume the padding is at most one page?)

Reply:
I do hope to completely remove mmap from JS though and implement this in terms of malloc+open+read.
Basically just copy https://github.com/WebAssembly/wasi-libc/blob/master/libc-bottom-half/mman/mman.c.

Reply:
#11576